### PR TITLE
fix(swap): size percentage presets by the asset's own precision

### LINKED ISSUE #4835: [Swap] Percentage buttons hardcode 4 decimals — 25% of a small BTC balance becomes 0; 100% strands dust (tests already specify the coin-aware rule)
## Problem

All swap percentage buttons (25/50/75/100) truncate the amount to a hardcoded **4 decimals**, regardless of the coin. Two user-visible consequences:

1. **Broken button on small native balances:** BTC balance `0.00021322` → 25% = `0.0000533` → `truncated(toPlaces: 4)` = `0.0000` → `fromAmount = "0"`. The button produces a zero-value swap (no quote / "amount too small") for any coin whose 25% share is below `0.0001` (BTC < ~0.0004, ETH < ~0.0004, all of WBTC…).
2. **100% strands dust:** WBTC balance `0.00021322` → "100%" swaps `0.0002`, stranding `0.00001322` (~$0.80 at $60k). A `0.00019` WBTC holder strands ~47% of their balance.

**Where:** `handlePercentageSelection` — [SwapDetailsScreen.swift#L460-L487](https://github.com/vultisig/vultisig-ios/blob/4d020d190/VultisigApp/VultisigApp/Features/Swap/Views/Screens/SwapDetailsScreen.swift#L460-L487), `let decimalsToUse: Int = 4` at L462, applied in every branch.

The intended coin-aware rule already exists in the tests: [SwapPercentageTests.swift#L96](https://github.com/vultisig/vultisig-ios/blob/4d020d190/VultisigApp/VultisigAppTests/Services/SwapPercentageTests.swift#L96) specifies `min(9, max(4, decimals))` — the UI just doesn't use it. (The Send flow uses `min(8, decimals)` and does not have this bug.)

## Suggested fix

Replace the hardcoded `4` with the coin-aware precision from the tests (or `min(8, decimals)` to match Send). For the 100% non-native case, use the raw balance directly instead of a decimal-truncated view.

## Related hardening

`Coin.raw(for:)` rounds **up** when converting a decimal amount to base units ([Coin.swift#L276-L281](https://github.com/vultisig/vultisig-ios/blob/4d020d190/VultisigApp/VultisigApp/Model/Coin.swift#L276-L281), `NSDecimalRound(..., .up)`), and `SwapPayloadBuilder` uses it for the on-chain amount. Today the 4-dp truncation masks it, but once full-precision amounts flow through (this fix, or manual full-precision entry equal to balance), rounding up can exceed the raw balance by 1 unit and fail on-chain. Truncate instead, or clamp to `rawBalance`.

Context: part of a cross-platform "perfect max on all assets" sweep (user report on Android of max USDT sends/swaps failing; Android and Windows sibling issues filed).

### PR DESCRIPTION
**Pull request by gastonm5**
Fixes #4835.

> **Stacked on `fix/send-max-precision-4836`.** This branch is cut from that
> branch, so the diff against `main` will show #4836's commit until it merges.
> Merge #4836 first; this PR's own change is the single commit on top.

## The bug

`handlePercentageSelection` hardcoded `let decimalsToUse: Int = 4` and applied it
to all five branches (25 / 50 / 75 / 100% native / 100% non-native):

- A BTC balance of `0.00021322` gave 25% = `0.000053305`, truncated to 4 places →
  **`"0"`**. The button produced a zero-value swap for any coin whose 25% share
  fell below `0.0001`.
- 100% non-native truncated to 4 places, stranding dust — an 8-decimal WBTC
  balance lost up to ~9 999 sats on every Max.

## The fix

One shared rule, `min(8, coin.decimals)`, in a new `PercentageAmountLogic`
(`Core/Utils/`), used by all three flows that had drifted apart:

| Call site | Before | Now |
|---|---|---|
| `SwapDetailsScreen.handlePercentageSelection` | hardcoded `4` | `SwapCryptoLogic.percentageAmountText` → shared helper |
| `LimitSwapFormViewModel.sourceAmountText` | hardcoded `4` (with a comment saying it deliberately matched the market buttons) | shared helper |
| `SendCryptoLogic.applyPercentage` | its own `coinDecimals > 8 ? 8 : coinDecimals` | shared helper (behavior unchanged) |

`min(9, max(4, decimals))` — the rule the old tests documented — was rejected:
the `max(4, …)` floor over-reports precision for assets with fewer than 4
decimals, producing amounts the asset cannot represent.

100% now converts the **raw balance** directly rather than a decimal-truncated
view, so it strands nothing. The native 100% still reserves the network fee.

The math was extracted out of the SwiftUI view into
`SwapCryptoLogic.percentageAmountText`, which is what makes it testable at all —
the view method is precisely why the old tests dodged it.

## Why the `raw(for:)` cap ships here

This is not unrelated hardening — **the precision fix arms a latent overflow**,
so the two cannot be separated.

`Coin.raw(for:)` rounds **UP**, and the amount field round-trips through
`String.toDecimal()`, which parses via a Double-backed `NumberFormatter` that
keeps only ~16 significant digits and rounds half-up. Verified:
`1.234567890123456789` parses back as `1.234567890123457`. So a 100% preset on an
18-decimal asset converted to **more base units than the wallet holds** and the
transaction would fail on chain. The old 4-decimal truncation masked this;
restoring full precision exposes it.

`SwapCryptoLogic.amountInCoinDecimal` therefore caps the result at `rawBalance`.
The cap is deliberately narrow and lives in the swap amount path, not inside
`raw(for:)` — that function is also used for UTXO/Cardano fee planning in
`SwapPayloadBuilder`, and the `.up` rounding mode is left untouched.

It cannot hide an over-spend: the insufficient-funds check is
`SwapCryptoLogic.balanceError`, which compares `Decimal` values and never calls
`amountInCoinDecimal` — that helper feeds only `SwapPayloadBuilder`. A test pins
this. A zero `rawBalance` (balance not yet loaded) is excluded, so a valid amount
is never capped to zero; `SwapCryptoLogicTests`' `amountInCoinDecimal("1.5", btc)
== 150_000_000` assertion is unaffected.

## Tests

`SwapPercentageTests` was **rewritten**. The old version never invoked production
code — it asserted `min`/`max` on literals plus `formatToDecimal`, and its own
header admitted as much. It passed with the bug present and would have passed
unchanged after the fix.

17 tests now exercise the extracted helpers: the reported BTC case at 25/50/75/100%
(25% asserted non-zero), 100% returning the exact full balance, 6-decimal USDC,
an 18-decimal token (8-place cap on fractions, full precision on 100%), a
sub-4-decimal asset, the fee reservation, and the cap — including that a
full-precision 18-decimal balance converts to exactly `rawBalance` and not one
unit more.

## Gate

`make test` on iPhone 16 Pro Max — `** TEST SUCCEEDED **`, **3138 tests, 1 skipped,
0 failures**. SwiftLint clean on all changed files.

Reviewed with the cross-model loop; Codex's final whole-branch pass returned
no findings (it caught a real missing-brace blocker one round earlier).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved max-send calculations to preserve full asset precision and avoid over-spending, including fee-adjusted caps and correct handling when balances are zero/unavailable.
  - Standardized percentage preset amount calculations across send and swap flows, with consistent truncation/rounding rules for low-decimal assets and fee scenarios.
  - Updated Cardano ADA validation/recommendation text to quote fee-adjusted “Send Max” with consistent 6-decimal precision.

- **Tests**
  - Expanded send-max, fee edge-case, locale-aware percentage, and Cardano “Send Max” suggestion/validation coverage to match the updated precision and formatting behavior.

### CODE REVIEW

File: VultisigApp/VultisigApp/Blockchain/Cardano/Signing/Cardano.swift [hunk 1]
```diff
@@ -98,10 +104,10 @@ class CardanoHelper {
         // 2. Check sufficient balance
         let totalNeeded = sendAmount + estimatedFee
         if totalBalance < totalNeeded {
-            // For MAX amount display, truncate to 5 decimal places to match getMaxValue behavior
+            // Quote the suggested MAX at ADA's full precision, matching getMaxValue.
             let totalBalanceDecimal = totalBalance.toADA
-            let truncatedBalance = totalBalanceDecimal.truncated(toPlaces: 5) // ADA has 6 decimals, so decimals - 1 = 5
-            let totalBalanceADA = truncatedBalance.formatToDecimal(digits: 5)
+            let truncatedBalance = totalBalanceDecimal.truncated(toPlaces: adaDisplayDecimals)
+            let totalBalanceADA = truncatedBalance.formatToDecimal(digits: adaDisplayDecimals)
```
[the remark this answers is not in the document]
Comment (gastonm5) [reply VultisigApp/VultisigApp/Blockchain/Cardano/Signing/Cardano.swift]: Fixed — this is the same finding CodeRabbit raised on #4848, which this branch stacks on, so it arrived here through the shared `Cardano.swift` diff rather than from this PR's own changes.

Addressed at source in #4848 (commit `d346d4988`): the three "Send Max" suggestion messages now derive the quoted amount from `totalBalance - estimatedFee` via a shared `sendMaxSuggestionADA` helper, matching `SendCryptoLogic.computeMaxAmount`, with `CardanoSendMaxSuggestionTests` covering all three paths plus the full-6-decimal and fee-exceeds-balance cases. Verified pre-existing on `origin/main` before fixing.

This branch has been rebased onto the corrected #4848 tip, so the fix is present here too. Re-gated after the rebase: `** TEST SUCCEEDED **`, 3143 executed / 1 skipped / 0 failures — 17 `SwapPercentageTests` and the 5 inherited `CardanoSendMaxSuggestionTests` all passing by name.

Resolving here; the substantive discussion is on #4848.

File: VultisigApp/VultisigApp/Features/Swap/ViewModels/SwapCryptoLogic.swift [hunk 2]
```diff
@@ -63,7 +63,47 @@ enum SwapCryptoLogic {
...
+        // never exceed the balance. Over-spends are caught elsewhere, on a
+        // separate `Decimal` path (`balanceError` compares `fromAmount` against
+        // `balanceDecimal` and never calls this), so capping here hides nothing
+        // from the insufficient-funds check.
+        //
+        // A zero `rawBalance` means the balance hasn't loaded yet rather than
+        // an empty wallet; capping to it would zero out a valid amount.
+        let rawBalance = fromCoin.rawBalance.toBigInt()
+        guard rawBalance > 0 else { return raw }
+        return min(raw, rawBalance)
```
[the remark this answers is not in the document]
Comment (gastonm5) [reply VultisigApp/VultisigApp/Features/Swap/ViewModels/SwapCryptoLogic.swift]: Rejecting this one — the premise is inverted, and I confirmed that the hard way.

I initially took the suggestion, made exactly the change proposed (`guard !rawBalance.isEmpty`), and the gate came back **`** TEST FAILED **`, 2 failures**:

- `SwapCryptoLogicTests.testAmountInCoinDecimalScalesByCoinDecimals` — expected `150000000`, got `0`
- `SwapPayloadBuilderTests.testThorchainPayloadRoutesThroughInboundAddressForNativeSource` — expected `100000000`, got `0`

Neither fixture sets `rawBalance` at all, which is what exposed the real behaviour:

```swift
// Model/Coin.swift — init
self.rawBalance = .zero          // String.zero == "0", not ""
```

And `rg rawBalance = ""` over `VultisigApp/` returns **zero** hits — nothing in the app ever assigns `""`. The `var rawBalance: String = ""` property default is unreachable, because `init` always overwrites it.

So it is the other way round: **`"0"` is the unloaded state.** Every coin reads `"0"` from construction until a balance lands, and "not yet loaded" is genuinely indistinguishable from "empty wallet" — there is no sentinel to branch on. Guarding on `!isEmpty` would cap to zero whenever the swap form renders before balances arrive, zeroing a valid amount. That is a live regression, which is what those two tests were catching.

Keeping `guard rawBalance > 0`. A genuinely empty wallet is already rejected upstream by `balanceError` on the separate `Decimal` path, so skipping the cap there hides nothing.

That said, the finding was useful: the old comment asserted the conclusion without the reason. `590e347c8` records why, and pins it with two tests — `testZeroRawBalanceDoesNotCapTheAmount` (must not cap) and `testFreshCoinSeedsRawBalanceToZeroString` (a fresh `Coin` carries `"0"`). If `Coin.init` ever gains a distinct unloaded sentinel, that second test fails and the distinction you are asking for becomes implementable. No production behaviour change.

Gate: `** TEST SUCCEEDED **`, 3145 executed / 1 skipped / 0 failures.

Leaving this open as a documented disagreement.

**Review by johnnyluo — APPROVED** [review]
LGTM
